### ISSUE TITLE
compiler segfaults with `void` and `distinct`

### ISSUE THREAD

**Issue by theSherwood** [nim-works/nimskull#1316]
### Specification

The compiler is segfaulting with some uses of `void`. In some cases, `void` can be used as a type argument to some generic type. In other cases it causes the compiler to segfault. I originally thought that `default` was in part to blame but I have no idea now.

### Example

```nim
type
  EntryBox*[K, V] = object
    entry*: tuple[key: K, value: V]

  Node*[K, V] = object
    entries*: array[2, EntryBox[K, V]]

  Wrapper1*[K, V] = object
    node*: Node[K, V]

  Wrapper2*[K] {.borrow: `.`.} = distinct Wrapper1[K, void]
  WrapperRef2*[K] = ref Wrapper2[K]

proc init_2*[K](): WrapperRef2[K]  =
  var res = default(WrapperRef2[K])
  res.node = Node[K, void]()
  return res

var s = init_2[int]()
```

### Actual Output

Fails to compile due to segfault:

```
SIGSEGV: Illegal storage access. (Attempt to read from nil?)
./scripts/build.sh: line 312: 58542 Segmentation fault: 11  ( ${NIM} -c --cc: ${CC} --gc: arc --opt: speed --threads: off --profiler: off --stackTrace: off --d: release --nimcache: ${PATH_TO_C_ASSETS} c ${FILE} )
```

### Expected Output

Should compile

### Additional Information

Similar to the snippet above, this snippet uses `void` and causes the compiler to segfault, despite the lack of use of `distinct`.

```nim
type
  EntryBox*[K, V] = object
    entry*: tuple[key: K, value: V]

  Node*[K, V] = object
    entries*: array[2, EntryBox[K, V]]

  Wrapper1*[K, V] = object
    node*: Node[K, V]
  WrapperRef1*[K, V] = ref Wrapper1[K, V]

  WrapperRef2*[K] = object
    re: WrapperRef1[K, void]

proc init_2*[K](): WrapperRef2[K]  =
  result.re = WrapperRef1[K, void]()

var s = init_2[int]()
```

The next snippet compiles without problem despite the use of `void` and prints `1`, which is what we would expect.

```nim
type
  Obj*[K, V] = object
    k*: K
    v*: V
  KeyOnly*[K] = Obj[K, void]

var ko = KeyOnly[uint8]()
echo sizeof(ko)
```

**Comment by zerbina** [comment]
Thank you for filing the issue!

The problem is with how generic tuple types are instantiated (`void` fields aren't removed properly). A minimal example reproducing the crash:
```nim
type Tup[T] = tuple[x: T]
var x: Tup[void]
echo sizeof(x)
```

**Comment by zerbina** [comment]
Fixed by #1319 -- the provided example now compiles successfully.
